File: src/dataset/typescript/binary-search_exercism-samajammin/claude/ClaudeSonnet4_binary-search_v3.ts
export default class BinarySearch {
  public array: number[]

  constructor(arr: number[]) {
    if (this.isSorted(arr)) {
      this.array = arr
    }
  }

  private isSorted(arr: number[]): boolean {
    for (let i = 1; i < arr.length; i++) {
      if (arr[i] < arr[i - 1]) {
        return false
      }
    }
    return true
  }

  private binarySearch(target: number, left: number, right: number): number {
    while (left <= right) {
      const midpoint = left + Math.floor((right - left) / 2)
      const value = this.array[midpoint]
      
      if (target === value) {
        return midpoint
      } else if (target > value) {
        left = midpoint + 1
      } else {
        right = midpoint - 1
      }
    }
    return -1
  }

  public indexOf(searchTerm: number): number {
    if (this.array === undefined) {
      throw new Error('Invalid array.')
    }
    return this.binarySearch(searchTerm, 0, this.array.length - 1)
  }
}

File: src/dataset/c/binary-search_exercism-yorange1/claude/ClaudeSonnet4_binary_search_v3.c
#include "binary_search.h"

int *binary_search(int value, const int *arr, size_t length)
{
    if (!arr || length == 0) return NULL;
    
    size_t left = 0;
    size_t right = length - 1;
    
    while (left <= right) {
        size_t mid = left + ((right - left) >> 1);
        int mid_val = arr[mid];
        
        if (mid_val == value) {
            return (int *)&arr[mid];
        }
        
        if (mid_val > value) {
            if (mid == 0) break;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    
    return NULL;
}